source/extensions/filters/listener/original_src/original_src_socket_option.cc

```diff
@@ -0,0 +1,70 @@
+#include "extensions/filters/listener/original_src/original_src_socket_option.h"
+
+#include <arpa/inet.h>
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace OriginalSrc {
+
+constexpr uint8_t OriginalSrcSocketOption::IPV4_KEY;
+constexpr uint8_t OriginalSrcSocketOption::IPV6_KEY;
+
+OriginalSrcSocketOption::OriginalSrcSocketOption(
+    Network::Address::InstanceConstSharedPtr src_address)
+    : src_address_(std::move(src_address)) {
+  // Source transparency only works on IP connections.
+  ASSERT(src_address_->type() == Network::Address::Type::Ip);
+}
+
+bool OriginalSrcSocketOption::setOption(
+    Network::Socket& socket, envoy::api::v2::core::SocketOption::SocketState state) const {
+
+  if (state == envoy::api::v2::core::SocketOption::STATE_PREBIND) {
+    socket.setLocalAddress(src_address_, false);
+  }
+
+  bool result = true;
+  std::for_each(options_to_apply_.begin(), options_to_apply_.end(),
+                [&socket, state](const Network::Socket::OptionConstSharedPtr& option) {
+                  option->setOption(socket, state);
```

Comment:
I plan on actually capturing the result as part of the for_each so I can fail the operation of one of them fails. I wanted to wait until I actually implemented the socket options before doing too much with it, though.